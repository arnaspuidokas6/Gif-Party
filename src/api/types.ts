export interface FetchGifsRequest {
    query?: string;
    limit?: string;
}

// https://developers.giphy.com/docs/api/schema/#image-object
export interface IGifResponse {
    imageUrl: string;
    title: string;
    userImage?: string;
    importedAt?: string;
    displayName?: string;
}

interface IUrl {
    url: string;
}

interface IImage {
    fixed_height: IUrl;
}

interface IUser {
    avatar_url?: string;
    display_name?: string;
}

// https://developers.giphy.com/docs/api/schema/#gif-object
export interface IGif {
    title?: string;
    images?: IImage;
    user?: IUser;
    import_datetime?: string;
}
