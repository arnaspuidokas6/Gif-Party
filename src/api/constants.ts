export const API_KEY = '5vXtKWdBdVUrfD74YUXuNvFVQ8OanFKS';
export const DEFAULT_LIST_ITEMS = 12;
export const VALIDATION_PATTERN = '[^A-Za-z0-9]';
