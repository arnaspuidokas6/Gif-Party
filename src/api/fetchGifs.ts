import axios from 'axios';
import { API_KEY } from './constants';
import { FetchGifsRequest, IGif, IGifResponse } from './types';

const createGif = (data: IGif): IGifResponse => {
    return {
        title: data.title ?? '',
        imageUrl: data?.images?.fixed_height.url ?? '',
        userImage: data?.user?.avatar_url ?? '',
        displayName: data?.user?.display_name ?? '',
    };
};

export const fetchGifs = async ({ query = 'happy', limit = '12' }: FetchGifsRequest): Promise<IGifResponse[]> => {
    const res: {
        data: { data: IGif[] };
    } = await axios.get(`https://api.giphy.com/v1/gifs/search?q=${query}&limit=${limit}&api_key=${API_KEY}`);

    return res.data.data.map(createGif);
};
