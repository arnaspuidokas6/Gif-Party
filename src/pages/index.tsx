import React, { createContext, Dispatch, FC, SetStateAction, useContext } from 'react';
import { IGifResponse } from '../api/types';

export const EMPTY_GIFS_LIST = [];
export const IS_VALID = true;
export const DEFAULT_SELECTED_ITEM = {} as IGifResponse;
export const DEFAULT_OPEN_MODAL = false;
const EMPTY_VOID = () => {};

interface IGifsContext {
    gifsList: IGifResponse[];
    selectedItem: IGifResponse;
    setSelectedItem: (value: IGifResponse) => void;
    openModal: boolean;
    setOpenModal: Dispatch<SetStateAction<boolean>>;
}

const GifsListContext = createContext<IGifsContext>({
    gifsList: EMPTY_GIFS_LIST,
    selectedItem: DEFAULT_SELECTED_ITEM,
    setSelectedItem: EMPTY_VOID,
    openModal: DEFAULT_OPEN_MODAL,
    setOpenModal: EMPTY_VOID,
});

export const useGifsContext = () => {
    const context = useContext(GifsListContext);
    if (!context) {
        throw new Error("This component can't be used outside GifsList component");
    }
    return context;
};

export const GifsListPage: FC<IGifsContext> = ({
    gifsList,
    selectedItem,
    setSelectedItem,
    children,
    openModal,
    setOpenModal,
}) => {
    const context: IGifsContext = {
        gifsList,
        selectedItem,
        setSelectedItem,
        setOpenModal,
        openModal,
    };
    return <GifsListContext.Provider value={context}>{children}</GifsListContext.Provider>;
};
