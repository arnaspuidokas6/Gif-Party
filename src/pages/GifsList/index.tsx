import React, { FC, useState, useEffect } from 'react';
import { fetchGifs } from '../../api/fetchGifs';
import { IGifResponse } from '../../api/types';
import { Card } from '../../components/Card';
import { Heading } from '../../components/Heading';
import { NoGifs } from '../../components/NoGifs';
import { SearchBar } from '../../components/SearchBar';
import '../../tailwind.output.css';

export const GifsList: FC = () => {
    const [gifsList, setGifsList] = useState<IGifResponse[]>();
    const [searchValue, setSearchValue] = useState<string>('');
    const [searchError, setSearchError] = useState<boolean>(false);

    const handleSearchChange = (event: { target: { value: string } }) => {
        const typedValue = event.target.value;
        const regexpPattern = '[^A-Za-z0-9]';

        if (typedValue.match(regexpPattern)) {
            setSearchError(true);
        } else {
            setSearchError(false);
            setSearchValue(typedValue);
        }
    };

    useEffect(() => {
        fetchGifs({ query: searchValue }).then((newGifsList) => setGifsList(newGifsList));
    }, [searchValue]);

    return (
        <div className="container mx-auto">
            <Heading />
            <SearchBar setSearchValue={handleSearchChange} isValid={!searchError} />
            <>
                {gifsList?.length ? (
                    <div className="mt-20 grid grid-cols-1 gap-8 md:grid-cols-2 xl:grid-cols-4">
                        {gifsList?.map((props, index) => (
                            <Card {...props} key={index} />
                        ))}
                    </div>
                ) : (
                    <NoGifs />
                )}
            </>
        </div>
    );
};
