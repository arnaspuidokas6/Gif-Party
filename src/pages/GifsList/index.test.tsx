import React from 'react';
import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DEFAULT_SELECTED_ITEM, EMPTY_GIFS_LIST, GifsListContext } from '../../pages';
import { IGifResponse } from '../../api/types';
import { GifsList } from '.';
import * as API from '../../api/fetchGifs';

const gifsList = EMPTY_GIFS_LIST as IGifResponse[];
const selectedItem = DEFAULT_SELECTED_ITEM;
const openModal = false;

const searchSpy = jest.spyOn(API, 'fetchGifs');
describe('Card', () => {
    const setOpenModal = jest.fn().mockResolvedValueOnce(true);
    const setSelectedItem = jest.fn().mockResolvedValueOnce(true);

    it('should match snapshot', async () => {
        const { container } = render(
            <GifsListContext.Provider value={{ gifsList, openModal, selectedItem, setSelectedItem, setOpenModal }}>
                <GifsList />
            </GifsListContext.Provider>,
        );
        expect(container).toMatchSnapshot();
    });

    it('should allow to call correct request and display list', async () => {
        const SEARCH_WORD = 'happy';
        const DEFAULT_LIMIT = '12';
        render(
            <GifsListContext.Provider value={{ gifsList, openModal, selectedItem, setSelectedItem, setOpenModal }}>
                <GifsList />
            </GifsListContext.Provider>,
        );

        act(() => {
            userEvent.type(screen.getByTestId('search-input'), SEARCH_WORD);
        });

        expect(searchSpy).toHaveBeenLastCalledWith({ limit: DEFAULT_LIMIT, query: SEARCH_WORD });
        await waitFor(() => expect(screen.queryByText('Sorry, no gifs :( try typing in search.')).toBeNull());
        expect(screen.getAllByTestId('gif-card')[11]).toBeVisible();
    });
});
