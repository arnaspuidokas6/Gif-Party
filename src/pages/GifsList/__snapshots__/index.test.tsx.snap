// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
<div>
  <div
    class="container mx-auto"
  >
    <div
      class="hero"
    >
      <div
        class="hero-headline flex flex-col items-center justify-center pt-24 text-center"
      >
        <h1
          class="font-bold text-3xl text-gray-900"
        >
          Wanna see some cool gifs? ü•∫
        </h1>
        <p
          class="font-base text-base text-gray-600"
        >
          use the search below and get them now üèÇ
        </p>
      </div>
    </div>
    <div
      class="box pt-6"
    >
      <div
        class="box-wrapper"
      >
        <div
          class="bg-white rounded flex items-center w-full p-3 shadow-sm border border-gray-200"
        >
          <input
            class="w-full pl-4 text-sm outline-none focus:outline-none bg-transparent"
            data-testid="search-input"
            name="query"
            placeholder="Search for gifs"
            type="search"
          />
        </div>
      </div>
    </div>
    <h1
      class="mt-20 font-bold text-3xl text-gray-900"
      data-testid="no-gifs"
    >
      Sorry, no gifs :( try typing in search.
    </h1>
  </div>
  <div
    class="hidden fixed bg-grayish w-full h-full top-0 left-0 bg-gray-200"
  >
    <div
      class="mx-auto px-4 py-8 max-w-xl my-20"
    >
      <div
        class="bg-white shadow-2xl rounded-xl mb-6 tracking-wide"
      >
        <div
          class="md:flex-shrink-0"
        >
          <img
            class="w-full h-full rounded-lg rounded-b-none"
          />
          <span
            class="absolute top-0 right-0 p-4 text-red-500"
          >
            <svg
              class="h-6 w-6 fill-current text-grey hover:text-grey-darkest"
              role="button"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M14.348 14.849a1.2 1.2 0 0 1-1.697 0L10 11.819l-2.651 3.029a1.2 1.2 0 1 1-1.697-1.697l2.758-3.15-2.759-3.152a1.2 1.2 0 1 1 1.697-1.697L10 8.183l2.651-3.031a1.2 1.2 0 1 1 1.697 1.697l-2.758 3.152 2.758 3.15a1.2 1.2 0 0 1 0 1.698z"
              />
            </svg>
          </span>
        </div>
        <div
          class="px-4 py-2"
        >
          <div
            class="author flex items-center -ml-3 my-3 mt-4"
          >
            <h2
              class="text-sm tracking-tighter text-gray-900"
            >
              <a
                href="#"
              >
                No user name
              </a>
               
              <span
                class="text-gray-600 pl-5"
              >
                No import data
              </span>
            </h2>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
